File: sdk/typescript/src/resources/secrets.ts
/**
 * Resource class for managing secrets via the Scion Hub API.
 *
 * @packageDocumentation
 */

import { BaseResource } from './base.js';
import type {
  Secret,
  SetSecretRequest,
  SetSecretResponse,
  SecretScopeOptions,
  ListSecretsOptions,
  ListSecretResponse,
} from '../types/secrets.js';

/** Result from listing secrets, including scope metadata. */
export interface SecretsPage {
  /** Secret metadata entries (no values). */
  data: Secret[];
  /** The scope these secrets belong to. */
  scope: string;
  /** The scope entity ID. */
  scopeId: string;
}

/**
 * Provides operations on Scion secrets.
 *
 * Secrets are scoped to a user, project, or runtime broker. Secret values
 * are write-only and are never returned by the API — only metadata is
 * accessible via list and get operations.
 *
 * @example
 * ```ts
 * const client = new ScionClient({ hubUrl: 'https://hub.example.com' });
 *
 * // List user-scoped secrets
 * const page = await client.secrets.list();
 *
 * // Set a secret
 * const result = await client.secrets.set('MY_API_KEY', {
 *   value: 'sk-...',
 *   description: 'API key for external service',
 * });
 * ```
 */
export class SecretsResource extends BaseResource {
  private static readonly BASE_PATH = '/api/v1/secrets';

  /**
   * List secret metadata for the specified scope.
   *
   * Returns metadata only — secret values are never included.
   *
   * @param params - Optional filtering parameters.
   * @returns A page of secret metadata.
   */
  async list(params?: ListSecretsOptions): Promise<SecretsPage> {
    const query = this.buildQuery({
      scope: params?.scope,
      scopeId: params?.scopeId,
      type: params?.type,
    });

    const response = await this.transport.request<ListSecretResponse>(
      'GET',
      SecretsResource.BASE_PATH,
      query ? { query } : undefined,
    );

    return {
      data: response.secrets ?? [],
      scope: response.scope,
      scopeId: response.scopeId,
    };
  }

  /**
   * Get metadata for a specific secret by key.
   *
   * The secret value is never returned.
   *
   * @param key - The secret key name.
   * @param params - Optional scope parameters.
   * @returns The secret metadata.
   * @throws {NotFoundError} If the secret is not found.
   */
  async get(key: string, params?: SecretScopeOptions): Promise<Secret> {
    const query = this.buildQuery({
      scope: params?.scope,
      scopeId: params?.scopeId,
    });

    return this.transport.request<Secret>(
      'GET',
      `${SecretsResource.BASE_PATH}/${encodeURIComponent(key)}`,
      query ? { query } : undefined,
    );
  }

  /**
   * Create or update a secret.
   *
   * @param key - The secret key name.
   * @param params - The secret parameters including value and optional metadata.
   * @returns The set secret response containing metadata and created flag.
   */
  async set(key: string, params: SetSecretRequest): Promise<SetSecretResponse> {
    return this.transport.request<SetSecretResponse>(
      'PUT',
      `${SecretsResource.BASE_PATH}/${encodeURIComponent(key)}`,
      { body: params },
    );
  }

  /**
   * Delete a secret by key.
   *
   * @param key - The secret key name.
   * @param params - Optional scope parameters.
   * @throws {NotFoundError} If the secret is not found.
   */
  async delete(key: string, params?: SecretScopeOptions): Promise<void> {
    const query = this.buildQuery({
      scope: params?.scope,
      scopeId: params?.scopeId,
    });

    await this.transport.request<void>(
      'DELETE',
      `${SecretsResource.BASE_PATH}/${encodeURIComponent(key)}`,
      query ? { query } : undefined,
    );
  }
}

File: sdk/python/src/scion/resources/secrets.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Secrets resource for the Scion Python SDK.

Provides synchronous and asynchronous access to the Hub secrets API.
Note: Secret values are write-only and never returned by the API.

Mirrors Go implementation in ``pkg/hubclient/secrets.go``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING
from urllib.parse import quote

from scion.types.secrets import (
    ListSecretResponse,
    Secret,
    SetSecretRequest,
    SetSecretResponse,
)

if TYPE_CHECKING:
    from scion._transport import AsyncTransport, Transport


def _scope_params(
    scope: str | None = None,
    scope_id: str | None = None,
) -> dict[str, str] | None:
    """Build query-parameter dict from optional scope fields.

    Returns ``None`` when both values are absent so the transport
    skips the ``params`` kwarg entirely.
    """
    params: dict[str, str] = {}
    if scope is not None:
        params["scope"] = scope
    if scope_id is not None:
        params["scopeId"] = scope_id
    return params or None


class SecretsResource:
    """Synchronous resource for managing secrets.

    Obtained via ``ScionClient.secrets`` -- not instantiated directly.

    Usage::

        client = ScionClient(base_url, token=token)

        # List secrets
        response = client.secrets.list(scope="project", scope_id="proj-123")

        # Get a single secret's metadata
        secret = client.secrets.get("MY_KEY")

        # Create or update a secret
        result = client.secrets.set("MY_KEY", "super-secret-value")

        # Delete a secret
        client.secrets.delete("MY_KEY")
    """

    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    def list(
        self,
        *,
        scope: str | None = None,
        scope_id: str | None = None,
    ) -> ListSecretResponse:
        """List secret metadata for the specified scope.

        Secret values are never returned by the API.

        Args:
            scope: Scope filter -- ``"user"``, ``"project"``, or
                ``"runtime_broker"``.  Defaults to the server's default
                (typically ``"user"``).
            scope_id: ID of the scoped entity.  Required when *scope*
                is ``"project"`` or ``"runtime_broker"``.

        Returns:
            A :class:`ListSecretResponse` containing secret metadata.
        """
        params = _scope_params(scope, scope_id)
        resp = self._transport.get("/api/v1/secrets", params=params)
        return ListSecretResponse.model_validate(resp.json())

    def get(
        self,
        key: str,
        *,
        scope: str | None = None,
        scope_id: str | None = None,
    ) -> Secret:
        """Get metadata for a specific secret by key.

        The secret value is never returned by the API.

        Args:
            key: The secret key to retrieve.
            scope: Scope filter (see :meth:`list`).
            scope_id: ID of the scoped entity (see :meth:`list`).

        Returns:
            A :class:`Secret` with metadata for the requested key.
        """
        params = _scope_params(scope, scope_id)
        path = f"/api/v1/secrets/{quote(key, safe='')}"
        resp = self._transport.get(path, params=params)
        return Secret.model_validate(resp.json())

    def set(
        self,
        key: str,
        value: str,
        *,
        scope: str | None = None,
        scope_id: str | None = None,
        description: str | None = None,
        injection_mode: str | None = None,
        secret_type: str | None = None,
        target: str | None = None,
        allow_progeny: bool | None = None,
    ) -> SetSecretResponse:
        """Create or update a secret.

        Args:
            key: The secret key.
            value: The secret value (write-only -- never returned).
            scope: Scope type -- ``"user"``, ``"project"``, or
                ``"runtime_broker"``.  Defaults to the server's default.
            scope_id: ID of the scoped entity.  Required when *scope*
                is ``"project"`` or ``"runtime_broker"``.
            description: Optional human-readable description.
            injection_mode: ``"always"`` or ``"as_needed"``
                (default: ``"as_needed"``).
            secret_type: Secret type -- ``"environment"`` (default),
                ``"variable"``, or ``"file"``.
            target: Projection target (defaults to *key*).
            allow_progeny: Allow creator's progeny agents to access
                the secret (user scope only).

        Returns:
            A :class:`SetSecretResponse` with the updated metadata and
            a ``created`` flag indicating whether the secret was new.
        """
        req = SetSecretRequest(
            value=value,
            scope=scope,
            scope_id=scope_id,
            description=description,
            injection_mode=injection_mode,
            secret_type=secret_type,
            target=target,
            allow_progeny=allow_progeny,
        )
        path = f"/api/v1/secrets/{quote(key, safe='')}"
        resp = self._transport.put(path, json=req.model_dump_api())
        return SetSecretResponse.model_validate(resp.json())

    def delete(
        self,
        key: str,
        *,
        scope: str | None = None,
        scope_id: str | None = None,
    ) -> None:
        """Delete a secret.

        Args:
            key: The secret key to delete.
            scope: Scope filter (see :meth:`list`).
            scope_id: ID of the scoped entity (see :meth:`list`).
        """
        params = _scope_params(scope, scope_id)
        path = f"/api/v1/secrets/{quote(key, safe='')}"
        self._transport.delete(path, params=params)


class AsyncSecretsResource:
    """Asynchronous resource for managing secrets.

    Obtained via ``AsyncScionClient.secrets`` -- not instantiated directly.

    Usage::

        client = AsyncScionClient(base_url, token=token)

        # List secrets
        response = await client.secrets.list(scope="project", scope_id="proj-123")

        # Get a single secret's metadata
        secret = await client.secrets.get("MY_KEY")

        # Create or update a secret
        result = await client.secrets.set("MY_KEY", "super-secret-value")

        # Delete a secret
        await client.secrets.delete("MY_KEY")
    """

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def list(
        self,
        *,
        scope: str | None = None,
        scope_id: str | None = None,
    ) -> ListSecretResponse:
        """List secret metadata for the specified scope.

        Secret values are never returned by the API.

        Args:
            scope: Scope filter -- ``"user"``, ``"project"``, or
                ``"runtime_broker"``.  Defaults to the server's default
                (typically ``"user"``).
            scope_id: ID of the scoped entity.  Required when *scope*
                is ``"project"`` or ``"runtime_broker"``.

        Returns:
            A :class:`ListSecretResponse` containing secret metadata.
        """
        params = _scope_params(scope, scope_id)
        resp = await self._transport.get("/api/v1/secrets", params=params)
        return ListSecretResponse.model_validate(resp.json())

    async def get(
        self,
        key: str,
        *,
        scope: str | None = None,
        scope_id: str | None = None,
    ) -> Secret:
        """Get metadata for a specific secret by key.

        The secret value is never returned by the API.

        Args:
            key: The secret key to retrieve.
            scope: Scope filter (see :meth:`list`).
            scope_id: ID of the scoped entity (see :meth:`list`).

        Returns:
            A :class:`Secret` with metadata for the requested key.
        """
        params = _scope_params(scope, scope_id)
        path = f"/api/v1/secrets/{quote(key, safe='')}"
        resp = await self._transport.get(path, params=params)
        return Secret.model_validate(resp.json())

    async def set(
        self,
        key: str,
        value: str,
        *,
        scope: str | None = None,
        scope_id: str | None = None,
        description: str | None = None,
        injection_mode: str | None = None,
        secret_type: str | None = None,
        target: str | None = None,
        allow_progeny: bool | None = None,
    ) -> SetSecretResponse:
        """Create or update a secret.

        Args:
            key: The secret key.
            value: The secret value (write-only -- never returned).
            scope: Scope type -- ``"user"``, ``"project"``, or
                ``"runtime_broker"``.  Defaults to the server's default.
            scope_id: ID of the scoped entity.  Required when *scope*
                is ``"project"`` or ``"runtime_broker"``.
            description: Optional human-readable description.
            injection_mode: ``"always"`` or ``"as_needed"``
                (default: ``"as_needed"``).
            secret_type: Secret type -- ``"environment"`` (default),
                ``"variable"``, or ``"file"``.
            target: Projection target (defaults to *key*).
            allow_progeny: Allow creator's progeny agents to access
                the secret (user scope only).

        Returns:
            A :class:`SetSecretResponse` with the updated metadata and
            a ``created`` flag indicating whether the secret was new.
        """
        req = SetSecretRequest(
            value=value,
            scope=scope,
            scope_id=scope_id,
            description=description,
            injection_mode=injection_mode,
            secret_type=secret_type,
            target=target,
            allow_progeny=allow_progeny,
        )
        path = f"/api/v1/secrets/{quote(key, safe='')}"
        resp = await self._transport.put(path, json=req.model_dump_api())
        return SetSecretResponse.model_validate(resp.json())

    async def delete(
        self,
        key: str,
        *,
        scope: str | None = None,
        scope_id: str | None = None,
    ) -> None:
        """Delete a secret.

        Args:
            key: The secret key to delete.
            scope: Scope filter (see :meth:`list`).
            scope_id: ID of the scoped entity (see :meth:`list`).
        """
        params = _scope_params(scope, scope_id)
        path = f"/api/v1/secrets/{quote(key, safe='')}"
        await self._transport.delete(path, params=params)
